src/microbe_stage/ProcessSystem.cs

```diff
@@ -377,81 +381,115 @@ private void ProcessNode(IProcessable processor, float delta, float inverseDelta
         }
 
         if (environmentModifier <= MathUtils.EPSILON)
-        {
             canDoProcess = false;
-        }
 
+        // Compute spaceConstraintModifier before updating the final use and input amounts
         foreach (var entry in processData.Inputs)
         {
             if (entry.Key.IsEnvironmental)
                 continue;
 
-            var inputRemoved = entry.Value * process.Rate * delta;
-
-            currentProcessStatistics?.AddInputAmount(entry.Key, inputRemoved * inverseDelta);
+            var inputRemoved = entry.Value * process.Rate * delta * environmentModifier * spaceConstraintModifier;
 
             // currentProcessStatistics?.AddInputAmount(entry.Key, 0);
+            // We don't multiply by delta here because we report the per-second values anyway. In the actual process
+            // output numbers (computed after testing the speed), we need to multiply by inverse delta
+            currentProcessStatistics?.AddInputAmount(entry.Key, entry.Value * process.Rate * environmentModifier);
 
-            // If not enough compound we can't do the process
-            if (bag.GetCompoundAmount(entry.Key) < inputRemoved)
+            // If not enough we can't run the process unless we can lower spaceConstraintModifier enough
+            var availableAmount = bag.GetCompoundAmount(entry.Key);
+            if (availableAmount < inputRemoved)
             {
-                canDoProcess = false;
-                currentProcessStatistics?.AddLimitingFactor(entry.Key);
+                bool canRun = false;
+
+                if (availableAmount > MathUtils.EPSILON)
+                {
+                    var neededModifier = availableAmount / inputRemoved;
+
+                    if (neededModifier > Constants.MINIMUM_RUNNABLE_PROCESS_FRACTION)
+                    {
+                        spaceConstraintModifier = neededModifier;
+                        canRun = true;
+
+                        // Due to rounding errors there can be very small disparity here between the amount available
+                        // and what we will take with the modifiers. See the comment in outputs for more details
+                    }
+                }
+
+                if (!canRun)
+                {
+                    canDoProcess = false;
+                    currentProcessStatistics?.AddLimitingFactor(entry.Key);
+                }
```

Comment:
```suggestion
                    var neededModifier = availableAmount / inputRemoved;

                    if (neededModifier > Constants.MINIMUM_RUNNABLE_PROCESS_FRACTION)
                    {
                        spaceConstraintModifier = neededModifier;
                        canRun = true;

                        // Due to rounding errors there can be very small disparity here between the amount available
                        // and what we will take with the modifiers. See the comment in outputs for more details
                    }
                    else
                    {
                        canDoProcess = false;
                        currentProcessStatistics?.AddLimitingFactor(entry.Key);
                    }
                }
```

I think you can dispense with the flag and one of the conditionals here.  If availableAmount is zero, neededModifier will be as well. This won't handle a negative MINIMUM_RUNNABLE_PROCESS_FRACTION, but I think it's safe to discount that.

Reply:
I wanted to early return if there is no compound to be had.
I wrote the logic the way I did because it was easier for me to reason about and make absolutely sure there isn't funny business going on if there is too little input compound.

Reply:
I don't see any return statements here.

I can't argue against it being easier to you, but in my experience nothing confuses a new dev to the codebase like flag variables, which is why I try to avoid them.

Reply:
I misspoke about the early return. I meant that I skip the entire logic block.

I think flag variables are used elsewhere, at least in my personal coding style I use them a lot.
Anyway the main point here is that this version I'm sure does the right thing. If I were to change this I would want to get at least a couple of people playtesting this extensively to ensure correctness.

Reply:
I guess it comes down to subjectivity then. I'd like to see this trimmed down, but I won't block over it.